// @openedx/frontend-build/example/dist/babel/App.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = App;
var _env = _interopRequireDefault(require("env.config"));
var _Image = _interopRequireDefault(require("./Image"));
var _apple = _interopRequireWildcard(require("./apple.svg"));
var _apple2 = _interopRequireDefault(require("./apple.jpg"));
require("./style.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
// eslint-disable-next-line react/function-component-definition
function App() {
  var newEnglandApples = ['macintosh', 'granny smith'];
  var allApples = [].concat(newEnglandApples, ['fuji', 'golden delicious']);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("h1", {
      children: "Test page"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "SCSS parsing tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h3", {
      children: "The Apples"
    }), " (\"The Apples\" should be red)", /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "ES6 parsing tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
      children: allApples.map(function (apple) {
        return /*#__PURE__*/(0, _jsxRuntime.jsx)("li", {
          children: apple
        }, apple);
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "JSX parsing tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_apple.ReactComponent, {
      style: {
        width: '10rem'
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "TSX parsing tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Image["default"], {
      src: _apple["default"],
      alt: "appleFromTsx",
      style: {
        width: '10rem'
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "Asset import tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
      src: _apple["default"],
      alt: "apple",
      style: {
        width: '10rem'
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
      src: _apple["default"],
      alt: "apple",
      style: {
        width: '10rem'
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("br", {}), /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
      src: _apple2["default"],
      alt: "apple",
      style: {
        width: '10rem'
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("p", {
      children: "Photo by Louis Hansel @shotsoflouis on Unsplash"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "process.env tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("p", {
      children: ["Test process.env variable: ", process.env.TEST_VARIABLE]
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("p", {
      children: ["Non-existent process.env variable (nothing here is good): ", process.env.I_AM_NOT_HERE]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "env.config.js tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("p", {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        children: "env.config.js boolean test: "
      }), _env["default"].FALSE_VALUE === false ? _env["default"].CORRECT_BOOL_VALUE : _env["default"].INCORRECT_BOOL_VALUE]
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("p", {
      children: ["env.config.js integer test: ", Number.isInteger(_env["default"].INTEGER_VALUE) ? 'It was an integer. Great.' : 'It was not an integer! Why not? ']
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      children: "Right-to-left language handling tests"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("p", {
      className: "text-align-right",
      children: "I'm aligned right, but left in RTL."
    })]
  });
}
