// @openedx/paragon/dist/TransitionReplace/index.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
import React from 'react';
import PropTypes from 'prop-types';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import classNames from 'classnames';
var TransitionReplace = /*#__PURE__*/function (_React$Component) {
  _inherits(TransitionReplace, _React$Component);
  var _super = _createSuper(TransitionReplace);
  function TransitionReplace(props) {
    var _this;
    _classCallCheck(this, TransitionReplace);
    _this = _super.call(this, props);
    _this.state = {
      height: null
    };
    _this.onChildEnter = _this.onChildEnter.bind(_assertThisInitialized(_this));
    _this.onChildEntering = _this.onChildEntering.bind(_assertThisInitialized(_this));
    _this.onChildEntered = _this.onChildEntered.bind(_assertThisInitialized(_this));
    _this.onChildExit = _this.onChildExit.bind(_assertThisInitialized(_this));
    _this.onChildExiting = _this.onChildExiting.bind(_assertThisInitialized(_this));
    _this.onChildExited = _this.onChildExited.bind(_assertThisInitialized(_this));
    return _this;
  }

  // Transition events are fired in this order:
  //
  // onEnter > onEntering > onEntered
  // onExit  > onExiting  > onExited
  //
  // Keep in mind that we always have two transitions happening
  // both the entering and leaving children
  //
  // We set the container height (for animation) in this order:
  //
  // 1. onChildExit         (explicitly set the height to match the current current)
  // 2. onChildEntering     (set the height to match the new content)
  // 3. onChildExited       (reset the height to null)
  _createClass(TransitionReplace, [{
    key: "onChildEnter",
    value: function onChildEnter(htmlNode) {
      if (this.props.onChildEnter) {
        this.props.onChildEnter(htmlNode);
      }
    }
  }, {
    key: "onChildEntering",
    value: function onChildEntering(htmlNode) {
      this.setState({
        height: htmlNode.offsetHeight
      });
      if (this.props.onChildEntering) {
        this.props.onChildEntering(htmlNode);
      }
    }
  }, {
    key: "onChildEntered",
    value: function onChildEntered(htmlNode) {
      this.setState({
        height: null
      });
      if (this.props.onChildEntered) {
        this.props.onChildEntered(htmlNode);
      }
    }
  }, {
    key: "onChildExit",
    value: function onChildExit(htmlNode) {
      this.setState({
        height: htmlNode.offsetHeight
      });
      if (this.props.onChildExit) {
        this.props.onChildExit(htmlNode);
      }
    }
  }, {
    key: "onChildExiting",
    value: function onChildExiting(htmlNode) {
      if (this.props.onChildExiting) {
        this.props.onChildExiting(htmlNode);
      }
    }
  }, {
    key: "onChildExited",
    value: function onChildExited(htmlNode) {
      this.setState({
        height: null
      });
      if (this.props.onChildExited) {
        this.props.onChildExited(htmlNode);
      }
    }
  }, {
    key: "renderChildTransition",
    value: function renderChildTransition(child) {
      var _this2 = this;
      if (!child.key && process.env.NODE_ENV === 'development') {
        // eslint-disable-next-line no-console
        console.warn("TransitionReplace: A child is missing a 'key' prop. Keys are required for any child of this component.");
      }
      // Makes the exiting and entering children occupy the same space
      // SCSS handles the crossfade so it can be easily overridden
      var commonChildStyles = {
        // Prevent margin-collapsing which throws off height calculations
        padding: '.1px 0'
      };
      var transitionStyles = {
        entering: {},
        entered: {},
        exiting: {
          position: 'absolute',
          top: 0,
          left: 0,
          right: 0,
          pointerEvents: 'none'
        },
        exited: {}
      };
      return /*#__PURE__*/React.createElement(CSSTransition, {
        timeout: {
          enter: this.props.enterDuration,
          exit: this.props.exitDuration
        },
        unmountOnExit: true,
        mountOnEnter: true,
        onEnter: this.onChildEnter,
        onEntering: this.onChildEntering,
        onEntered: this.onChildEntered,
        onExit: this.onChildExit,
        onExiting: this.onChildExiting,
        onExited: this.onChildExited,
        classNames: this.props.transitionClassNames
      }, function (state) {
        return /*#__PURE__*/React.createElement("div", {
          style: _objectSpread(_objectSpread(_objectSpread({}, commonChildStyles), transitionStyles[state]), _this2.props.transitionStyles[state])
        }, child);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(TransitionGroup, {
        className: classNames('pgn-transition-replace-group', 'position-relative', {
          'overflow-hidden': this.state.height !== null
        }, this.props.className),
        style: {
          height: this.state.height
        }
      }, React.Children.map(this.props.children, this.renderChildTransition, this));
    }
  }]);
  return TransitionReplace;
}(React.Component);
TransitionReplace.propTypes = {
  /** Specifies an additional class for the base element */
  children: PropTypes.element,
  /** Duration of the element appearance transition. */
  enterDuration: PropTypes.number,
  /** Duration of the element dismiss transition. */
  exitDuration: PropTypes.number,
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** A `Transition` callback fired immediately after the `enter` or `appear` class is applied. */
  onChildEnter: PropTypes.func,
  /** A `Transition` callback fired immediately after the `enter-active` or `appear-active` class is applied. */
  onChildEntering: PropTypes.func,
  /**
   * A `Transition` callback fired immediately after the `enter` or
   * `appear` classes are removed and the done class is added to the DOM node.
   */
  onChildEntered: PropTypes.func,
  /** A `Transition` callback fired immediately after the `exit` class is applied. */
  onChildExit: PropTypes.func,
  /** A `Transition` callback fired immediately after the `exit-active` is applied. */
  onChildExiting: PropTypes.func,
  /**
   * A `Transition` callback fired immediately after the `exit` classes
   * are removed and the exit-done class is added to the DOM node.
   */
  onChildExited: PropTypes.func,
  /** An object that specifies transition styles. */
  transitionStyles: PropTypes.shape({
    entering: PropTypes.shape({}),
    entered: PropTypes.shape({}),
    exiting: PropTypes.shape({}),
    exited: PropTypes.shape({})
  }),
  /** Specifies class name to append to the `Transition`. */
  transitionClassNames: PropTypes.string
};
TransitionReplace.defaultProps = {
  children: undefined,
  enterDuration: 300,
  exitDuration: 300,
  className: undefined,
  onChildEnter: undefined,
  onChildEntering: undefined,
  onChildEntered: undefined,
  onChildExit: undefined,
  onChildExiting: undefined,
  onChildExited: undefined,
  transitionStyles: {},
  transitionClassNames: 'pgn__transition-replace'
};
export default TransitionReplace;
