// @openedx/paragon/dist/Stack/index.js
var _excluded = ["direction", "gap", "reversed", "children", "className"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
var DIRECTION_VARIANTS = ['horizontal', 'vertical'];
var Stack = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var direction = _ref.direction,
    gap = _ref.gap,
    reversed = _ref.reversed,
    children = _ref.children,
    className = _ref.className,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: classNames(direction === 'horizontal' ? 'pgn__hstack' : 'pgn__vstack', gap ? "pgn__stack-gap--".concat(gap) : '', reversed ? 'pgn__stack-reversed' : '', className)
  }, rest), children);
});
Stack.propTypes = {
  /** Specifies the content of the `Stack`. */
  children: PropTypes.node.isRequired,
  /** Specifies direction of the children blocks (column/row). */
  direction: PropTypes.oneOf(DIRECTION_VARIANTS),
  /**
   * Specifies inner space between children blocks.
   *
   * Valid values are based on `the spacing classes`:
   * `0, 0.5, ... 6`.
   */
  gap: PropTypes.number,
  /** Specifies the order of the children. */
  reversed: PropTypes.bool,
  /** Specifies an additional `className` to add to the base element. */
  className: PropTypes.string
};
Stack.defaultProps = {
  direction: 'vertical',
  gap: 0,
  className: undefined,
  reversed: false
};
export default Stack;
