// @openedx/paragon/dist/Dropzone/DefaultContent.js
import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, useIntl } from 'react-intl';
import { getTypesString, formatBytes } from './utils';
import messages from './messages';
import Icon from '../Icon';
import { FileUpload } from '../../icons';
function DefaultContent(_ref) {
  var accept = _ref.accept,
    minSize = _ref.minSize,
    maxSize = _ref.maxSize;
  var intl = useIntl();
  var getFileRestrictionMessage = function getFileRestrictionMessage() {
    var fileTypePart;
    var fileSizePart;
    if (accept) {
      var allTypes = getTypesString(accept);
      var lastTypeSeparatorLocation = allTypes.lastIndexOf(',');
      fileTypePart = intl.formatMessage(messages.fileTypeRestriction, {
        count: lastTypeSeparatorLocation === -1 ? 1 : 2,
        firstPart: lastTypeSeparatorLocation === -1 ? allTypes : allTypes.slice(0, lastTypeSeparatorLocation),
        secondPart: lastTypeSeparatorLocation !== -1 && allTypes.slice(lastTypeSeparatorLocation + 1)
      });
    }
    if (minSize && maxSize && Number.isFinite(maxSize)) {
      fileSizePart = intl.formatMessage(messages.fileSizeBetween, {
        sizeMin: formatBytes(minSize),
        sizeMax: formatBytes(maxSize)
      });
    } else if (maxSize && Number.isFinite(maxSize)) {
      fileSizePart = intl.formatMessage(messages.fileSizeMax, {
        sizeMax: formatBytes(maxSize)
      });
    } else if (minSize) {
      fileSizePart = intl.formatMessage(messages.fileSizeMin, {
        sizeMin: formatBytes(minSize)
      });
    }
    if (fileTypePart && fileSizePart) {
      return "".concat(fileTypePart, " (").concat(fileSizePart, ")");
    }
    if (fileTypePart) {
      return fileTypePart;
    }
    return fileSizePart;
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    className: "pgn__dropzone__upload-icon-wrapper"
  }, /*#__PURE__*/React.createElement(Icon, {
    src: FileUpload,
    className: "pgn__dropzone__upload-icon"
  })), /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement(FormattedMessage, {
    id: "pgn.Dropzone.DefaultContent.label",
    description: "A text that appears as a label for input of Dropzone component.",
    defaultMessage: "Drag and drop your file here or click to upload."
  })), [accept, minSize, maxSize].some(function (value) {
    return value;
  }) && /*#__PURE__*/React.createElement("p", {
    className: "pgn__dropzone__upload-restriction-message"
  }, getFileRestrictionMessage()));
}
DefaultContent.propTypes = {
  accept: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.string)),
  maxSize: PropTypes.number,
  minSize: PropTypes.number
};
DefaultContent.defaultProps = {
  accept: undefined,
  maxSize: undefined,
  minSize: undefined
};
export default DefaultContent;
