// @openedx/paragon/dist/Dropzone/UploadProgress.js
import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Button from '../Button';
import ProgressBar from '../ProgressBar';
import Spinner from '../Spinner';
function UploadProgress(_ref) {
  var percent = _ref.percent,
    variant = _ref.variant,
    name = _ref.name,
    onCancel = _ref.onCancel;
  if (variant === 'spinner') {
    return /*#__PURE__*/React.createElement(Spinner, {
      animation: "border",
      "aria-live": "polite",
      screenReaderText: "Uploading ".concat(name, ", ").concat(percent, "% done."),
      "data-testid": "upload-spinner"
    });
  }
  return /*#__PURE__*/React.createElement("div", {
    className: "w-50"
  }, /*#__PURE__*/React.createElement("p", {
    className: "text-muted"
  }, /*#__PURE__*/React.createElement(FormattedMessage, {
    id: "pgn.Dropzone.UploadProgress.uploadLabel",
    defaultMessage: "Uploading {filename}.",
    description: "A text that is shown near a progress bar during file upload in Dropzone component.",
    values: {
      filename: name
    }
  })), /*#__PURE__*/React.createElement("div", {
    className: "d-flex justify-content-between align-items-center w-100"
  }, /*#__PURE__*/React.createElement(ProgressBar, {
    now: percent,
    label: "".concat(percent, "%"),
    variant: "success",
    className: "flex-grow-1",
    "data-testid": "upload-progress-bar"
  }), /*#__PURE__*/React.createElement(Button, {
    variant: "tertiary",
    className: "ml-3",
    onClick: onCancel
  }, /*#__PURE__*/React.createElement(FormattedMessage, {
    id: "pgn.Dropzone.UploadProgress.cancelLabel",
    defaultMessage: "Cancel",
    description: "Label of a cancel button that is shown during file upload in Dropzone component."
  }))));
}
UploadProgress.propTypes = {
  variant: PropTypes.oneOf(['spinner', 'bar']).isRequired,
  percent: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onCancel: PropTypes.func.isRequired
};
export default UploadProgress;
