// @openedx/paragon/dist/DataTable/selection/ControlledSelectionStatus.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import React, { useContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import DataTableContext from '../DataTableContext';
import BaseSelectionStatus from './BaseSelectionStatus';
import { clearSelectionAction, setSelectAllRowsAllPagesAction, setSelectedRowsAction } from './data/actions';
import { getUnselectedPageRows, getRowIds } from './data/helpers';
function ControlledSelectionStatus(_ref) {
  var className = _ref.className,
    clearSelectionText = _ref.clearSelectionText;
  var _useContext = useContext(DataTableContext),
    itemCount = _useContext.itemCount,
    page = _useContext.page,
    _useContext$controlle = _slicedToArray(_useContext.controlledTableSelections, 2),
    _useContext$controlle2 = _useContext$controlle[0],
    selectedRows = _useContext$controlle2.selectedRows,
    isEntireTableSelected = _useContext$controlle2.isEntireTableSelected,
    dispatch = _useContext$controlle[1];
  useEffect(function () {
    if (isEntireTableSelected) {
      var selectedRowIds = getRowIds(selectedRows);
      var unselectedPageRows = getUnselectedPageRows(selectedRowIds, page);
      if (unselectedPageRows.length) {
        dispatch(setSelectedRowsAction(unselectedPageRows, itemCount));
      }
    }
  }, [isEntireTableSelected, selectedRows, itemCount, page, dispatch]);
  var numSelectedRows = isEntireTableSelected ? itemCount : selectedRows.length;
  var numSelectedRowsOnPage = (page || []).filter(function (r) {
    return r.isSelected;
  }).length;
  var selectionStatusProps = {
    className: className,
    numSelectedRows: numSelectedRows,
    numSelectedRowsOnPage: numSelectedRowsOnPage,
    clearSelectionText: clearSelectionText,
    onSelectAll: function onSelectAll() {
      return dispatch(setSelectAllRowsAllPagesAction());
    },
    onClear: function onClear() {
      return dispatch(clearSelectionAction());
    }
  };
  return /*#__PURE__*/React.createElement(BaseSelectionStatus, selectionStatusProps);
}
ControlledSelectionStatus.defaultProps = {
  className: undefined,
  clearSelectionText: undefined
};
ControlledSelectionStatus.propTypes = {
  /** A class name to append to the base element */
  className: PropTypes.string,
  /** A text that appears on the `Clear selection` button, defaults to 'Clear Selection' */
  clearSelectionText: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
};
export default ControlledSelectionStatus;
