// @openedx/paragon/dist/DataTable/utils/getVisibleColumns.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { useMemo, useContext, useEffect } from 'react';
import { CheckboxControl } from '../../Form';
import DataTableContext from '../DataTableContext';
import useConvertIndeterminateProp from './useConvertIndeterminateProp';
export var selectColumn = {
  id: 'selection',
  // The header can use the table's getToggleAllPageRowsSelectedProps or getToggleAllRowsSelectedProps
  // method to render a checkbox. The method is determined based on whether pagination is enabled or
  // not (i.e., ``page`` is defined).
  // Proptypes disabled as these props are passed in separately
  /* eslint-disable-next-line react/prop-types */
  Header: function Header(_ref) {
    var getToggleAllPageRowsSelectedProps = _ref.getToggleAllPageRowsSelectedProps,
      getToggleAllRowsSelectedProps = _ref.getToggleAllRowsSelectedProps,
      page = _ref.page;
    var _useContext = useContext(DataTableContext),
      isSelectable = _useContext.isSelectable,
      maxSelectedRows = _useContext.maxSelectedRows;
    var toggleRowsSelectedProps = useMemo(function () {
      // determine if this selection is for an individual page or the entire table
      var getToggleRowsSelectedProps = page ? getToggleAllPageRowsSelectedProps : getToggleAllRowsSelectedProps;
      return getToggleRowsSelectedProps();
    }, [getToggleAllPageRowsSelectedProps, getToggleAllRowsSelectedProps, page]);
    var updatedProps = useConvertIndeterminateProp(toggleRowsSelectedProps);
    var formatMaxSelectedRows = Math.max(0, maxSelectedRows);
    if (isSelectable && formatMaxSelectedRows >= 0) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      className: "pgn__data-table__controlled-select"
    }, /*#__PURE__*/React.createElement(CheckboxControl, _extends({}, updatedProps, {
      "data-testid": "datatable-select-column-checkbox-header"
    })));
  },
  // The cell can use the individual row's getToggleRowSelectedProps method
  // to the render a checkbox
  // Proptypes disabled as this prop is passed in separately
  /* eslint-disable react/prop-types */
  Cell: function Cell(_ref2) {
    var _selectedRowsOrdered;
    var row = _ref2.row;
    var _useContext2 = useContext(DataTableContext),
      isSelectable = _useContext2.isSelectable,
      maxSelectedRows = _useContext2.maxSelectedRows,
      onMaxSelectedRows = _useContext2.onMaxSelectedRows,
      _useContext2$state = _useContext2.state,
      selectedRowIds = _useContext2$state.selectedRowIds,
      selectedRowsOrdered = _useContext2$state.selectedRowsOrdered;
    var updatedProps = useConvertIndeterminateProp(row.getToggleRowSelectedProps());
    var index = row.index;
    var isRowSelected = (index in selectedRowIds);
    var selectedRowsLength = Object.keys(selectedRowIds).length;
    var formatMaxSelectedRows = Math.max(0, maxSelectedRows);
    var hasMaxSelectedRows = formatMaxSelectedRows === selectedRowsLength;
    var disableCheck = isSelectable && hasMaxSelectedRows && !isRowSelected;
    var lastRowSelected = (_selectedRowsOrdered = selectedRowsOrdered === null || selectedRowsOrdered === void 0 ? void 0 : selectedRowsOrdered[selectedRowsOrdered.length - 1]) !== null && _selectedRowsOrdered !== void 0 ? _selectedRowsOrdered : null;
    useEffect(function () {
      if (hasMaxSelectedRows && lastRowSelected === index) {
        onMaxSelectedRows === null || onMaxSelectedRows === void 0 ? void 0 : onMaxSelectedRows();
      }
    }, [hasMaxSelectedRows, index, isRowSelected, lastRowSelected, onMaxSelectedRows, selectedRowIds]);
    return /*#__PURE__*/React.createElement("div", {
      className: "pgn__data-table__controlled-select"
    }, /*#__PURE__*/React.createElement(CheckboxControl, _extends({}, updatedProps, {
      disabled: disableCheck,
      "data-testid": "datatable-select-column-checkbox-cell"
    })));
  },
  /* eslint-enable react/prop-types */
  disableSortBy: true
};
var getVisibleColumns = function getVisibleColumns(isSelectable, visibleColumns) {
  var additionalColumns = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var manualSelectColumn = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : selectColumn;
  var columns = [];
  if (isSelectable) {
    columns.push(manualSelectColumn);
  }
  columns = columns.concat(visibleColumns);
  if (additionalColumns.length > 0) {
    columns = columns.concat(additionalColumns);
  }
  return columns;
};
export default getVisibleColumns;
