// @openedx/paragon/dist/Tabs/index.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var _excluded = ["children", "className", "moreTabText", "defaultActiveKey", "activeKey"],
  _excluded2 = ["title", "notification", "tabClassName"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import React, { useEffect, useMemo, useRef, useState, useCallback } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import BaseTabs from 'react-bootstrap/Tabs';
import TabsDeprecated from './deprecated';
import Bubble from '../Bubble';
import Dropdown from '../Dropdown';
import useIndexOfLastVisibleChild from '../hooks/useIndexOfLastVisibleChild';
import Tab from './Tab';
export var MORE_TAB_TEXT = 'More...';
function Tabs(_ref) {
  var _overflowElementRef$c, _overflowElementRef$c3;
  var children = _ref.children,
    className = _ref.className,
    _ref$moreTabText = _ref.moreTabText,
    moreTabText = _ref$moreTabText === void 0 ? MORE_TAB_TEXT : _ref$moreTabText,
    defaultActiveKey = _ref.defaultActiveKey,
    activeKey = _ref.activeKey,
    props = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState(null),
    _useState2 = _slicedToArray(_useState, 2),
    containerElementRef = _useState2[0],
    setContainerElementRef = _useState2[1];
  var overflowElementRef = useRef(null);
  var indexOfLastVisibleChild = useIndexOfLastVisibleChild(containerElementRef === null || containerElementRef === void 0 ? void 0 : containerElementRef.firstChild, (_overflowElementRef$c = overflowElementRef.current) === null || _overflowElementRef$c === void 0 ? void 0 : _overflowElementRef$c.parentNode);
  useEffect(function () {
    if (containerElementRef) {
      var observer = new MutationObserver(function (mutations) {
        mutations.forEach(function (mutation) {
          // React-Bootstrap attribute 'data-rb-event-key' is responsible for the tab identification
          var eventKey = mutation.target.getAttribute('data-rb-event-key');
          // React-Bootstrap attribute 'aria-selected' is responsible for selected/unselected state
          var isActive = mutation.target.getAttribute('aria-selected') === 'true';
          // datakey attribute is added manually to the dropdown
          // elements so that they correspond to the native tabs' eventKey
          var element = containerElementRef.querySelector("[datakey='".concat(eventKey, "']"));
          var moreTab = containerElementRef.querySelector('.pgn__tab_more');
          if (isActive) {
            element === null || element === void 0 ? void 0 : element.classList.add('active');
            // Here we add active class to the 'More Tab' if element exists in the dropdown
            if (element) {
              moreTab.classList.add('active');
            } else {
              moreTab.classList.remove('active');
            }
          } else {
            element === null || element === void 0 ? void 0 : element.classList.remove('active');
          }
        });
      });
      observer.observe(containerElementRef, {
        attributes: true,
        subtree: true,
        attributeFilter: ['aria-selected']
      });
      return function () {
        return observer.disconnect();
      };
    }
    return undefined;
  }, [containerElementRef]);
  useEffect(function () {
    var _overflowElementRef$c2;
    if ((_overflowElementRef$c2 = overflowElementRef.current) !== null && _overflowElementRef$c2 !== void 0 && _overflowElementRef$c2.parentNode) {
      overflowElementRef.current.parentNode.tabIndex = -1;
    }
  }, [(_overflowElementRef$c3 = overflowElementRef.current) === null || _overflowElementRef$c3 === void 0 ? void 0 : _overflowElementRef$c3.parentNode]);
  var handleDropdownTabClick = useCallback(function (eventKey) {
    var hiddenTab = containerElementRef.querySelector("[data-rb-event-key='".concat(eventKey, "']"));
    hiddenTab.click();
  }, [containerElementRef]);
  var tabsChildren = useMemo(function () {
    var indexOfOverflowStart = indexOfLastVisibleChild + 1;
    var handleDropdownKeyPress = function handleDropdownKeyPress(e, eventKey) {
      if (e.key === 'Enter') {
        handleDropdownTabClick(eventKey);
      }
    };
    var childrenList = React.Children.map(children, function (child, index) {
      var _child$type;
      if ((child === null || child === void 0 ? void 0 : (_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.name) !== 'Tab' && process.env.NODE_ENV === 'development') {
        // eslint-disable-next-line no-console
        console.error("Tabs children can only be of type Tab. ".concat(children[index], " was passed instead."));
      }
      if (! /*#__PURE__*/React.isValidElement(child)) {
        return child;
      }
      var _child$props = child.props,
        title = _child$props.title,
        notification = _child$props.notification,
        tabClassName = _child$props.tabClassName,
        rest = _objectWithoutProperties(_child$props, _excluded2);
      var newTitle;
      if (notification) {
        newTitle = /*#__PURE__*/React.createElement(React.Fragment, null, title, /*#__PURE__*/React.createElement(Bubble, {
          variant: "error",
          role: "status",
          className: "pgn__tab-notification",
          "aria-live": "polite",
          expandable: true
        }, notification));
      } else {
        newTitle = title;
      }
      var tabClass = index > indexOfLastVisibleChild ? 'pgn__tab_invisible' : '';
      var modifiedTab = /*#__PURE__*/React.cloneElement(child, _objectSpread(_objectSpread({}, rest), {}, {
        title: newTitle,
        tabClassName: classNames(tabClass, tabClassName)
      }));
      return modifiedTab;
    });
    var moreTabHasNotification = false;
    var overflowChildren = childrenList.slice(indexOfOverflowStart).map(function (overflowChild) {
      if (!moreTabHasNotification && overflowChild.props.notification) {
        moreTabHasNotification = true;
      }
      return /*#__PURE__*/React.createElement(Dropdown.Item, {
        as: "li",
        tabIndex: "0",
        key: "".concat(overflowChild.props.eventKey, "overflow"),
        onClick: function onClick() {
          return handleDropdownTabClick(overflowChild.props.eventKey);
        },
        onKeyPress: function onKeyPress(e) {
          return handleDropdownKeyPress(e, overflowChild.props.eventKey);
        },
        disabled: overflowChild.props.disabled,
        datakey: overflowChild.props.eventKey,
        className: classNames({
          active: overflowChild.props.eventKey === defaultActiveKey || overflowChild.props.eventKey === activeKey
        }, 'pgn__tabs__dropdown-item')
      }, overflowChild.props.title);
    });
    childrenList.splice(indexOfOverflowStart, 0, /*#__PURE__*/React.createElement(Tab, {
      key: "moreTabKey",
      tabClassName: classNames(!overflowChildren.length && 'pgn__tab_invisible', 'pgn__tab_more'),
      title: /*#__PURE__*/React.createElement(Dropdown, {
        ref: overflowElementRef
      }, /*#__PURE__*/React.createElement(Dropdown.Toggle, {
        variant: "link",
        className: "nav-link",
        id: "pgn__tab-toggle"
      }, moreTabText, moreTabHasNotification && /*#__PURE__*/React.createElement(Bubble, {
        variant: "error",
        role: "status",
        className: "pgn__tab-notification"
      })), /*#__PURE__*/React.createElement(Dropdown.Menu, {
        as: "ul",
        className: "dropdown-menu-right"
      }, overflowChildren))
    }));
    return childrenList;
  }, [activeKey, children, defaultActiveKey, indexOfLastVisibleChild, moreTabText, handleDropdownTabClick]);
  return /*#__PURE__*/React.createElement("div", {
    ref: setContainerElementRef
  }, /*#__PURE__*/React.createElement(BaseTabs, _extends({
    defaultActiveKey: defaultActiveKey,
    activeKey: activeKey
  }, props, {
    className: classNames(className, 'pgn__tabs')
  }), tabsChildren));
}
Tabs.propTypes = {
  /** Specifies variant to use. */
  variant: PropTypes.oneOf(['tabs', 'pills', 'inverse-tabs', 'inverse-pills', 'button-group']),
  /** Specifies elements that is processed to create tabs. */
  children: PropTypes.node.isRequired,
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** Specifies text for the 'More' tab. */
  moreTabText: PropTypes.string,
  /** Specifies default active tab (uncontrolled usage). */
  defaultActiveKey: PropTypes.string,
  /** Specifies active tab (controlled usage). */
  activeKey: PropTypes.string
};
Tabs.defaultProps = {
  variant: 'tabs',
  className: undefined,
  moreTabText: MORE_TAB_TEXT,
  defaultActiveKey: undefined,
  activeKey: undefined
};
Tabs.Deprecated = TabsDeprecated;
export default Tabs;
export { Tab };
export { default as TabContainer } from 'react-bootstrap/TabContainer';
export { default as TabContent } from 'react-bootstrap/TabContent';
export { default as TabPane } from 'react-bootstrap/TabPane';
